### PR TITLE
pprint dynamo optimized model

### LINKED ISSUE #1691: printing an optimized model is not pretty
Eager mode print looks like this:

```python
>>> print(model)

Net(
  (fc1): Linear(in_features=10, out_features=20, bias=True)
  (fc2): Linear(in_features=20, out_features=30, bias=True)
)
```

Optimized looks like this
```python
>>> print(dynamo.optimize()(model))
<torch._dynamo.eval_frame._TorchDynamoContext.__call__.<locals>.TorchDynamoNNModuleWrapper object at 0x7f155ba76580
```

### PR DESCRIPTION
**Pull request by msaroufim**
Fixes https://github.com/pytorch/torchdynamo/issues/1691

Hopefully just adding `__repr__` is less controversial than shallow copying the `nn.Module` we can do thatt in a followup PR potentialyl


## Test

```
(sourcetorch) ubuntu@ip-172-31-31-152:~/pytorch$ pytest test/dynamo/test_misc.py 
=============================================================================================================== test session starts ===============================================================================================================
platform linux -- Python 3.8.13, pytest-7.2.0, pluggy-1.0.0
rootdir: /home/ubuntu/pytorch, configfile: pytest.ini
collected 136 items                                                                                                                                                                                                                               

test/dynamo/test_misc.py .s....STAGE:2022-11-02 21:32:01 13183:13183 ActivityProfilerController.cpp:294] Completed Stage: Warm Up
STAGE:2022-11-02 21:32:01 13183:13183 ActivityProfilerController.cpp:300] Completed Stage: Collection
......................................................s.........................................................STAGE:2022-11-02 21:32:03 13183:13183 ActivityProfilerController.cpp:294] Completed Stage: Warm Up
STAGE:2022-11-02 21:32:03 13183:13183 ActivityProfilerController.cpp:300] Completed Stage: Collection
..................                                                                           [100%]

========================================================================================================= 134 passed, 2 skipped in 7.42s ==========================================================================================================

```

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

### CODE REVIEW

**Comment by Chillee** [comment]
Can you do a more exhaustive comparison of this API? For example, can you verify that we don't increase memory usage by copying the module? And that torchdynamo actually works? 

Also, cc: @albanD on whether shallow copy of the module is going to have any unexpected effects.

File: torch/_dynamo/eval_frame.py [hunk 1]
```diff
@@ -120,27 +119,9 @@ def __call__(self, fn):
...
-                def forward(self, *args, **kwargs):
-                    return optimized_forward(*args, **kwargs)
-
-                def __call__(self, *args, **kwargs):
-                    return self.forward(*args, **kwargs)
-
-            new_mod = TorchDynamoNNModuleWrapper()
-            # Save the function pointer to find the original callable while nesting
-            # of decorators.
-            new_mod._torchdynamo_orig_callable = mod
```
Comment (anijain2305) [inline torch/_dynamo/eval_frame.py:143]: You forgot to keep this line of code. This is useful for nested torchdynamo optimize calls.

File: torch/_dynamo/eval_frame.py [hunk 2]
```diff
@@ -120,27 +119,9 @@ def __call__(self, fn):
         if isinstance(fn, torch.nn.Module):
             mod = fn
             optimized_forward = self(mod.forward)
+            new_mod = copy.copy(fn)
```
[the remark this answers is not in the document]
Comment (msaroufim) [reply torch/_dynamo/eval_frame.py:122]: We could do that but i was wondering/hoping there'd be other benefits to making an optimized model an nn.module outside of just printing

File: torch/_dynamo/eval_frame.py [hunk 3]
```diff
@@ -150,13 +153,20 @@ def __call__(self, *args, **kwargs):
 
         @functools.wraps(fn)
         def _fn(*args, **kwargs):
-            if is_fx_tracing():
+            any_arg_is_proxy = any(
```
Comment (msaroufim) [inline torch/_dynamo/eval_frame.py:156]: This was a contribution by `lintrunner -a torch/_dynamo/eval_frame.py`

File: torch/_dynamo/eval_frame.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Chillee) [reply torch/_dynamo/eval_frame.py]: @msaroufim I don't think `__repr__` is enough. A lot of the other things folks want to do with modules are not handled by monkeypatching the `__repr__`.

@albanD do you have any concrete objections to doing a shallow copy?

File: torch/_dynamo/eval_frame.py [hunk 4]
```diff
@@ -137,6 +137,9 @@ def forward(self, *args, **kwargs):
                 def __call__(self, *args, **kwargs):
                     return self.forward(*args, **kwargs)
 
+                def __repr__(self, *args, **kwargs):
+                    return self.__repr__(*args, **kwargs)
```
Comment (albanD) [inline torch/_dynamo/eval_frame.py:141]: Isn't that infinite recursing? you meant `mod.__repr__()` ?

**Review by albanD — COMMENTED** [review]
You might want a small smoke test to make sure it runs?

**Review by albanD — APPROVED** [review]
SGTM !

**Comment by Chillee** [comment]
No, I disagree with merging this.

**Review by Chillee — CHANGES_REQUESTED** [review]
@albanD what's the issue with the shallow copy?

**Comment by msaroufim** [comment]
Closing since it seems like we instead have went for a module wrapper

**Pull request closed without merging**